### PR TITLE
Tag directive for federation 2 support

### LINKED ISSUE #5240: Support for community directives - `tag` for Apollo contracts
### Is your feature request related to a problem?

There are some published directives that specific providers might support - in this case, looking at the `tag` directive to support [Apollo Contracts](https://www.apollographql.com/docs/studio/contracts/)

Opening this issue to see whether there's support for including general directives in this library? It isn't a GraphQL standard (that I'm aware of) so is kind of provider specific. But equally can be related to Apollo Federation.

### The solution you'd like

I have a local implementation, but would like to submit it as a PR to be available in the HotChocolate library if maintainers are open to this happening. And equally guidance on which project/namespace it'd sit.

### Product

Hot Chocolate

### PR DESCRIPTION
**Pull request by damienpontifex**
- Implementation of [`@tag`](https://www.apollographql.com/docs/federation/federated-types/federated-directives/#tag) directive as defined in federation 2 spec
- Largely a copy of the `@provides` setup, tests etc as base for work to provide this `@tag` directive

Closes #5240

### CODE REVIEW

**Comment by damienpontifex** [comment]
@michaelstaib any feedback on this? Or next steps?

**Comment by damienpontifex** [comment]
@michaelstaib I added in `TagDirective` type such that it's now `TagDirectiveType : DirectiveType<TagDirective>` for the use case of type-safe declaration as described on https://chillicream.com/docs/hotchocolate/v13/defining-a-schema/directives#usage-within-types when using code-first approach.

I noticed though that other directives don't have this approach, so let me know feedback on whether this seems supported, and recommended to add, or to revert to align with the others?

**Comment by michaelstaib** [comment]
Hey @damienpontifex,

sorry for letting you hang here :) 

We will start merging things again in around two weeks...

At the moment we are fully focused on version 14 goals but will switch back to version 13 and hope to get this in by May 2nd.

File: src/HotChocolate/ApolloFederation/src/ApolloFederation/TagDirectiveType.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+/// extend type User @key(fields: "id") {
+///   id: ID! @external
+///   email: String @tag(name: "public")
+///   customerNotes: String @tag(name: "internal")
+/// }
+/// </example>
+/// </summary>
+public sealed class TagDirective
+{
+    public string Name { get; set; } = default!;
```
Comment (michaelstaib) [inline src/HotChocolate/ApolloFederation/src/ApolloFederation/TagDirectiveType.cs:23]: This should be immutable and use constructor injection of the value

**Comment by michaelstaib** [comment]
@damienpontifex this looks good ... I have one minor issue ... the TagDirective should look like the following:

```
public sealed class TagDirective
{
    public TagDirective(string name)
    {
         Name = name ?? throw new Argument....;
    } 

    public string Name { get; }
}
```

File: src/HotChocolate/ApolloFederation/src/ApolloFederation/TagDirectiveType.cs (hunk 1 above)
Comment (damienpontifex) [reply src/HotChocolate/ApolloFederation/src/ApolloFederation/TagDirectiveType.cs]: @michaelstaib updated as per your suggestion in comment below to make it immutable and value taken from constructor

**Review by michaelstaib — APPROVED** [review]

**Review by michaelstaib — CHANGES_REQUESTED** [review]
Found one more issue ... 

The test `AnnotateTagToClassAttributePureCodeFirst` fails ... 

I have updated the snapshot tests but this one was checking that tag is annotated so not just an invalid snapshot.

**Comment by michaelstaib** [comment]
@damienpontifex can you check whats wrong with the test and fix it? If not for this test I would have merged it today.

**Comment by damienpontifex** [comment]
> @damienpontifex can you check whats wrong with the test and fix it? If not for this test I would have merged it today.

My initial investigation is it might be because `TagAttribute` inherits from `ObjectFieldDescriptorAttribute` so it isn't rendered onto the types. Investigating/changing now to see if I can figure that out to fix

**Comment by damienpontifex** [comment]
@michaelstaib I believe I've resolved that particular one. Please see latest commit for changes associated

**Comment by damienpontifex** [comment]
@michaelstaib any help I can provide to move this forward? 
Just noticed the push back on milestone to an unspecified 13 release, so offering any assistance I can to help progress this if possible

**Comment by damienpontifex** [comment]
I just saw #6341 and it seems like that provides everything we want/need that it'd also be usable as a directive for federation

i.e. `@tag(name: "your-value")`

**Comment by damienpontifex** [comment]
Actually doesn't seem to quite fit. Tried it out and some incompatible definition of the directive
```
DIRECTIVE_DEFINITION_INVALID: Invalid definition for directive "@tag": argument "name" should have type "String!" but found type "String"
    DIRECTIVE_DEFINITION_INVALID: Invalid definition for directive "@tag": "@tag" should have locations FIELD_DEFINITION, OBJECT, INTERFACE, UNION, ARGUMENT_DEFINITION, SCALAR, ENUM, ENUM_VALUE, INPUT_OBJECT, INPUT_FIELD_DEFINITION, but found (non-subset) SCHEMA, SCALAR, OBJECT, FIELD_DEFINITION, ARGUMENT_DEFINITION, INTERFACE, UNION, ENUM, ENUM_VALUE, INPUT_OBJECT, INPUT_FIELD_DEFINITION
```

**Comment by damienpontifex** [comment]
@michaelstaib is this something you're still open to having merged? And if so, any guide on milestone target we can work towards? I've seen the work on fusion within HotChocolate, so just gauging timelines or likelihood for any Apollo federation continuation.

**Comment by michaelstaib** [comment]
I am closing this pr... tag is coming with 13.6 and I should have a preview of it today

**Comment by michaelstaib** [comment]
Hey @damienpontifex we have now fixed the issues you pointed out with:
https://github.com/ChilliCream/graphql-platform/releases/tag/13.6.0-preview.31

Have fun... if there still are issues please raise an issue.

**Comment by damienpontifex** [comment]
Brilliant, thank you @michaelstaib

**Comment by damienpontifex** [comment]
Might also want to close/link to #5240 as the original issue raised for this support

**Pull request closed without merging**